pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -220,28 +247,21 @@ func (c *DeploymentConfigController) findDetails(config *deployapi.DeploymentCon
 		// Check if the image stream tag pointed by the trigger exists
 		if _, err := c.osClient.ImageStreamTags(config.Namespace).Get(name, tag); err != nil {
 			if !errors.IsNotFound(err) {
-				glog.V(2).Infof("Error while trying to get image stream tag %q: %v", istag, err)
-				return details, existingDeployments, latestDeploymentExists, nil
+				return "", fmt.Errorf("couldn't get image stream tag %q: %v", istag, err)
 			}
 			// In case the image stream tag was not found, then it either doesn't exist or doesn't exist yet
 			// (a build configuration output points to it so it's going to be populated at some point in the
 			// future)
-			details = fmt.Sprintf("The image trigger for image stream tag %q will have no effect because image stream tag %q does not exist.", istag, istag)
-			bcList, err := c.osClient.BuildConfigs(kapi.NamespaceAll).List(labels.Everything(), fields.Everything())
-			if err != nil {
-				glog.V(2).Infof("Error while trying to list build configs: %v", err)
-				return details, existingDeployments, latestDeploymentExists, nil
-			}
-
-			for _, bc := range bcList.Items {
+			invalidIsTags = append(invalidIsTags, istag)
```

Comment:
"Image stream tag foo:bar does not exist, a build may not have completed or
the tag may not be correct".

To make this message slightly more accurate you are holding a copy of all
builds configs in memory (1k per build config, and number of build configs
per project averages 1).  Say 10M for a 10k project cluster.  You then have
to copy it out of the store (so allocate ~25% of the total memory use as
garbage) and iterate over it.  2.5M garbage per check (probably) That makes
this O(NM) for N deployment configs and M build configs, which is effective
O(N^2).  It's not really a terrible check because the constant factor will
be small, but it is a lot of cost for a minimal reward.  In the long run,
it would be better to have this question answered by an API that can
efficiently answer the question "what builds push to this image stream".



On Nov 10, 2015, at 8:23 AM, Dan Mace <notifications@github.com> wrote:

In pkg/deploy/controller/deploymentconfig/controller.go
<https://github.com/openshift/origin/pull/5530#discussion_r44427514>:

>  			}
>  			// In case the image stream tag was not found, then it either doesn't exist or doesn't exist yet
>  			// (a build configuration output points to it so it's going to be populated at some point in the
>  			// future)
> -			details = fmt.Sprintf("The image trigger for image stream tag %q will have no effect because image stream tag %q does not exist.", istag, istag)
> -			bcList, err := c.osClient.BuildConfigs(kapi.NamespaceAll).List(labels.Everything(), fields.Everything())
> -			if err != nil {
> -				glog.V(2).Infof("Error while trying to list build configs: %v", err)
> -				return details, existingDeployments, latestDeploymentExists, nil
> -			}
> -
> -			for _, bc := range bcList.Items {
> +			invalidIsTags = append(invalidIsTags, istag)

I don't have any data, but anecdotally based on my interactions with users,
the top reason by far for unexpectedly missing deployments is because of
builds failing or not occurring. I think that both the CLI and the web
console need to connect those dots for users. We know the CLI can just use
the graph lib to report this on the fly. Where does that leave the console?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5530/files#r44427514>.